File: apps/android-customer/feature/compare/src/main/java/co/zw/nissangtr/customer/compare/GuestCompareStore.kt
package co.zw.nissangtr.customer.compare

import android.content.Context
import co.zw.nissangtr.customer.rpc.CompareItem
import co.zw.nissangtr.customer.rpc.RpcNames

/**
 * Guest compare tray — SharedPreferences OEM list (web `compare-selection` localStorage parity).
 * Auth users sync via list/add/remove compare RPCs.
 */
object GuestCompareStore {
    private const val PREFS = "gtr_customer_compare"
    private const val KEY = "gtr.compare.oems"

    fun readOems(context: Context): List<String> {
        val raw = prefs(context).getString(KEY, null)
            ?.split('\u001e')
            ?.filter { it.isNotBlank() }
            ?: emptyList()
        return normalize(raw)
    }

    fun writeOems(context: Context, oems: List<String>): List<String> {
        val next = normalize(oems).take(RpcNames.MAX_COMPARE_ITEMS)
        prefs(context).edit().putString(KEY, next.joinToString("\u001e")).apply()
        return next
    }

    fun addOem(context: Context, oem: String): List<String> {
        val needle = oem.trim()
        require(needle.isNotEmpty()) { "OEM required." }
        val list = readOems(context).toMutableList()
        if (list.any { it.equals(needle, ignoreCase = true) }) return list
        require(list.size < RpcNames.MAX_COMPARE_ITEMS) {
            "Compare holds up to ${RpcNames.MAX_COMPARE_ITEMS} SKUs. Remove one first."
        }
        list.add(0, needle)
        return writeOems(context, list)
    }

    fun removeOem(context: Context, oem: String): List<String> {
        val needle = oem.trim()
        val next = readOems(context).filter { !it.equals(needle, ignoreCase = true) }
        return writeOems(context, next)
    }

    fun clear(context: Context) {
        prefs(context).edit().remove(KEY).apply()
    }

    /** Map guest OEMs into [CompareItem] rows (synthetic ids) for matrix UI. */
    fun asCompareItems(context: Context): List<CompareItem> =
        readOems(context).mapIndexed { idx, oem ->
            CompareItem(
                id = "00000000-0000-4000-8000-%012d".format(idx + 1),
                stockItemId = "00000000-0000-4000-8000-%012d".format(idx + 100),
                oemPartNumber = oem,
                description = null,
                createdAt = null,
            )
        }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun normalize(oems: List<String>): List<String> {
        val seen = mutableSetOf<String>()
        val out = mutableListOf<String>()
        for (raw in oems) {
            val oem = raw.trim()
            if (oem.isEmpty()) continue
            val key = oem.lowercase()
            if (!seen.add(key)) continue
            out.add(oem)
        }
        return out
    }
}

File: apps/ios/Sources/GTRCustomerCore/GuestCompareStore.swift
import Foundation

/// Guest compare tray — UserDefaults OEM list (web `compare-selection` localStorage parity).
/// Auth users sync via `list_customer_compare_items` / add / remove RPCs.
@MainActor
public enum GuestCompareStore {
    private static let key = "gtr.compare.oems"

    public static func readOems() -> [String] {
        let raw = UserDefaults.standard.stringArray(forKey: key) ?? []
        return normalize(raw)
    }

    @discardableResult
    public static func writeOems(_ oems: [String]) -> [String] {
        let next = normalize(oems).prefix(maxCompareItems).map { $0 }
        UserDefaults.standard.set(Array(next), forKey: key)
        return Array(next)
    }

    public static func addOem(_ oem: String) throws -> [String] {
        let needle = oem.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !needle.isEmpty else {
            throw StorefrontError.message("OEM required.")
        }
        var list = readOems()
        if list.contains(where: { $0.caseInsensitiveCompare(needle) == .orderedSame }) {
            return list
        }
        guard list.count < maxCompareItems else {
            throw StorefrontError.message("Compare holds up to \(maxCompareItems) SKUs. Remove one first.")
        }
        list.insert(needle, at: 0)
        return writeOems(list)
    }

    public static func removeOem(_ oem: String) -> [String] {
        let needle = oem.trimmingCharacters(in: .whitespacesAndNewlines)
        let next = readOems().filter { $0.caseInsensitiveCompare(needle) != .orderedSame }
        return writeOems(next)
    }

    public static func clear() {
        UserDefaults.standard.removeObject(forKey: key)
    }

    /// Map guest OEMs into `CompareItem` rows (synthetic ids) for matrix UI.
    public static func asCompareItems() -> [CompareItem] {
        readOems().enumerated().map { idx, oem in
            CompareItem(
                id: UUID(uuidString: String(format: "00000000-0000-4000-8000-%012d", idx + 1))
                    ?? UUID(),
                stockItemId: UUID(),
                oemPartNumber: oem,
                description: nil,
                createdAt: nil
            )
        }
    }

    private static func normalize(_ oems: [String]) -> [String] {
        var seen = Set<String>()
        var out: [String] = []
        for raw in oems {
            let oem = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !oem.isEmpty else { continue }
            let key = oem.lowercased()
            guard !seen.contains(key) else { continue }
            seen.insert(key)
            out.append(oem)
        }
        return out
    }
}
